### ISSUE TITLE
Please document (lack of) equality guarantees for ImmutableMap.values()

### ISSUE THREAD

**Issue by malcolmr** [google/guava#3255]
`ImmutableMap` already documents that `values()` iterates the values in the order that entries were added to the map. However, the return type is `ImmutableCollection`, which does not have any particular equality guarantees per se (it inherits the non-guarantees from `Collection`).

I would have assumed that for any two `ImmutableMap` instances that contained equal values added in the same order, `values()` would return collections that would compare equal. However, this does not appear to be guaranteed (and appears to definitely be incorrect for some implementations, like `ImmutableMapValues`, which have no `equals()` method).

Assuming no guarantees are intended, would it be reasonable to add a note to `values()` pointing out that  — unlike `entrySet()` and `keySet()` —  there is no guarantee that the collection returned from `values()` has a sensible equality implementation (and perhaps pointing at `values().asList()` as a (usually cheap) way to get a collection that does).

(I guess this applies to the `values()` methods in `ImmutableMultimap`, `ImmutableSortedMap`, and `ImmutableTable` too?)

**Comment by kevinb9n** [comment]
Technically this applies to every method that returns `Collection`, `Iterable`, `Multimap`, `Queue`, `CharSequence`, or any other type without well-defined equality. Which is a lot of methods, so (a) we have [Error Prone](http://errorprone.info) checkers to report equals() reliance on one of these types, and (b) maybe we can figure out some rule for when the warning is called for and when it isn't on these methods...?

**Comment by kevinb9n** [comment]
Also when `ImmutableCollection` dissuades you from relying on `equals` I like the idea of plugging the `asList` view right there.

**Comment by cpovirk** [comment]
+1 to Kevin.

Now, I have seen problems from this specific misunderstanding before, so maybe that's enough to justify it on such a common interface. It would be great if this were emphasized on `java.util.Map.values()`, since the same problem exists for _all_ maps :\ But even if it were, I could see how people could expect even more of `ImmutableMap`, since it makes the additional promise of ordering.

(Maybe we should learn from that lesson and put such docs on `Multimap.values()` and `entries()`?)

Another tricky bit is that certain `values()` and `entries()` implementations return `Set` and thus _do_ have content-based `equals()`. And strangely, it can happen even if you don't explicitly use, say, `ImmutableBiMap`: `ImmutableMap.of("one key", "one value").values()` is a `Set` -- but only in the `-jre` version of Guava, not the `-android` version! Maybe all this subtlety is alone enough reason to justify more docs: Most methods that return `Collection` (not that there are many, or at least there _shouldn't_ be...) presumably aren't being overridden to return `Set` in some cases.

**Comment by malcolmr** [comment]
Good point about ErrorProne (this would be https://errorprone.info/bugpattern/UndefinedEquals, I assume?).

Worth noting that that check presently won't trigger for `ImmutableCollection`, since it requires the receiver type to be exactly `Collection` (or `Iterable`, etc). I guess we could add it easily enough, though?

And yes, the "mostly this will work except where it doesn't" aspect was the part that was troubling me here.

Would it be worth adding `@Deprecated ImmutableCollection.equals()` with a pointer to `asList()`? (Or is that only warranted for methods like `ImmutableList.add()` that are always an error?)

**Comment by Meijuh** [comment]
I would like to note the importance of solving this issue. The following simple code already fails, without clear documentation why:

    final ArrayListMultimap<String, String> alm1 = ArrayListMultimap.create();
    final ArrayListMultimap<String, String> alm2 = ArrayListMultimap.create();
    assert alm1.values().equals(alm2.values());
